internal/proxyservice/proxy_service.go

```diff
@@ -1,43 +1,147 @@
-	// TODO: Implement
+type connectorResources struct {
+	logger logapi.Logger
+	config []byte
+}
+
+func (cr *connectorResources) Logger() logapi.Logger {
+	return cr.logger
+}
+
+func (cr *connectorResources) Config() []byte {
+	return cr.config
+}
+
+func newConnectorResources(l logapi.Logger, cfg []byte) connector.Resources {
+	return &connectorResources{
+		logger: l,
+		config: cfg,
+	}
+}
 
-	// For each ProxyService:
-	//
-	// 1. Rewrap the Logger with service name prefix
-	// 2. Create the ConnectorResources object
-	// etc...
+// Start starts all proxy services
+func (s *proxyServices) Start() error {
+	for _, svc := range s.servicesToStart() {
+		err := svc.Start()
+		if err != nil {
+			s.logger.Errorf("could not start proxy service: %s", err)
+			continue
+		}
+		s.runningServices = append(s.runningServices, svc)
+	}
+	return nil
 }
 
 // Stop stops all proxy services
-func (s *proxyServices) Stop() {
-	// Stop all the Service
+func (s *proxyServices) Stop() error {
+	for _, svc := range s.runningServices {
+		err := svc.Stop()
+		if err != nil {
+			s.logger.Errorf("could not stop proxy service: %s", err)
+		}
+	}
+	return nil
+}
+
+func (s *proxyServices) servicesToStart() []secretless.Service {
+	var servicesToStart []secretless.Service
+
+	tcpPlugins := s.availPlugins.TCPPlugins()
+	// httpPlugins := s.availPlugins.HTTPPlugins()
+
+	for _, svc := range s.config.Services {
+		requestedPlugin := svc.Connector //TODO: this rename is a name smell
+
+		// first check the available TCP Plugins
+		tcpPlugin, found := tcpPlugins[requestedPlugin]
+		if found {
+			if tcpSvc := s.createTCPService(svc, tcpPlugin); tcpSvc != nil {
+				servicesToStart = append(servicesToStart, tcpSvc)
+			}
+			continue
+		}
+
+		// TODO: next check available HTTP Plugins
+		// httpPlugin, found := httpPlugins[requestedPlugin]
+
+		// TODO: Deal with SSH in a hardcoded way
+
+		// Default case: not found
+		s.logger.Errorf("plugin '%s' not available.", requestedPlugin)
+	}
+	return servicesToStart
+}
+
+func (s *proxyServices) createTCPService(
+	svc *v2.Service,
+	plugin tcp.Plugin,
+) secretless.Service {
+
+	//TODO: Add validation somewhere about overlapping listenOns
+	listener, err := net.Listen("tcp", strings.TrimLeft(svc.ListenOn, "tcp://"))
+	if err != nil {
+		// TODO: Should we do more than this?
+		s.logger.Errorf("could not create listener on: %s", svc.ListenOn)
+		return nil
+	}
+
+	svcLogger := s.logger.CopyWith(svc.Name, s.logger.DebugEnabled())
+	connResources := newConnectorResources(svcLogger, svc.ConnectorConfig)
+	connector_ := plugin.NewConnector(connResources)
+
+	// Temp var required so that the function closes over the current
+	// loop value.
+	credsCopy := svc.Credentials
+	credsRetriever := func() (map[string][]byte, error) {
+		return GetSecrets(credsCopy)
+	}
+
+	newSvc, err := tcpproxy.NewProxyService(
+		connector_,
+		listener,
+		svcLogger,
+		credsRetriever,
+	)
+
+	if err != nil {
+		// TODO: Should we do more than this?
```

Comment:
Yes, this is also a terminal error I think

Reply:
Addressed by solution to previous comment